src/ngraph/runtime/gpu/cuda_emitter.cpp

```diff
@@ -280,6 +314,295 @@ size_t runtime::gpu::CUDAEmitter::build_1d_max_pool(const GPURuntimeContext* ctx
+                writer << "int mt = m * str_d - pad_d;\n";
+
+                writer << "int pool_size = ";
+                auto pool_size = include_pad ? "TRS" : "0";
+                writer << pool_size << ";\n";
+
+                writer << "float sum = 0.0f;\n";
+                writer << "float rcp_pool_size = 1.0f;\n";
+                // each warp operates on a single pooling window and
+                // reduces the contents of the window within the warp
+                writer << "for (int trs = tid; trs < TRS; trs += 32)\n";
+                writer.block_begin();
+                {
+                    writer << "int t = div64(trs, magic_RS, shift_RS);\n";
+                    writer << "int rs = mod16(trs, t, RS);\n";
+                    writer << "int r  = div64(rs, magic_S, shift_S);\n";
+                    writer << "int s  = mod16(rs, r, S);\n";
+
+                    // coordinate transformation from TRS to DHW
+                    // via MPQ transform factors above
+                    writer << "int x = qs + s;\n";
+                    writer << "int y = pr + r;\n";
+                    writer << "int z = mt + t;\n";
+
+                    // helper to check participating threads
+                    writer << "bool bounds_x = (x >= 0) && (x < W);\n";
+                    writer << "bool bounds_y = (y >= 0) && (y < H);\n";
+                    writer << "bool bounds_z = (z >= 0) && (z < D);\n";
+                    writer << "bool within_tensor_bounds = bounds_x && bounds_y && bounds_z;\n";
+
+                    if (include_pad == false)
+                    {
+                        // count the number of (non-padded) elements
+                        writer << "pool_size += __popc(__ballot(within_tensor_bounds));\n";
+                    }
+                    // this will need to change to k->c once
+                    // feature pooling support is added
+                    writer << "int idx = n*CDHW + k*DHW + z*HW + y*W + x;\n";
+                    writer << "sum += load(in,idx,within_tensor_bounds);\n";
+                }
+                writer.block_end();
+
+                writer << "rcp_pool_size = 1.0f / (float)pool_size;\n";
+                // reduce pooling window within warp.
+                // this could be improved by calculating the
+                // pooling windows each thread can partake in to
+                // reduce loads and increase coalescing. in that case,
+                // multiple warps per block would be required and the
+                // warp reduced sums would need to be accumulated in
+                // shared memory
+                writer << "for (int i = 16; i > 0; i >>= 1)\n";
+                writer.block_begin();
+                {
+                    writer << "sum += __shfl_xor(sum,i);\n";
+                }
+                writer.block_end();
+                // write result to output
+                writer << "if (tid == 0)\n";
```

Comment:
since you only need tid 0 for result, have you try __shfl_down.

Reply:
I may be wrong on this but isn't shuffle_down just: 
```reg(tid) += register(tid+1),```
and so to accumulate all 32 threads I would need to call __shfl_down 32 times.

Since __shfl_xor uses a butterfly merge, each half of the selected threads are added to the other half, and this is done recursively, so only 4 calls to __shfl_xor are needed.

Reply:
so far as I know, shuffle down is similar to shuffle xor, but can only go in one direction, so the result will only be stored in tid = 0; something like.
reg(tid) += reg(tid+n); (n>0)
it will be less work than _shfl_xor,  `n + n/2 + n/4 +..`  vs `n + n + n + ...`

Reply:
https://devblogs.nvidia.com/faster-parallel-reductions-kepler/

Reply:
__shfl_down(reg, 16, 32);
__shfl_down(reg, 8, 16);
__shfl_down(reg, 4, 8);
...

Reply:
for _shfl_xor, it should be the case. 
for my understanding,  _shfl_down should not be. let me check.

Reply:
Correction: In shuffle_down the values are not broadcast. They threads out of the range just return their local value.

Ah maybe I see. Exchanging data between part of the warp vs the whole warp should still cost the same number of instructions (down vs xor), but with down, you have branch divergence in the warp, but you also don't have to wait for the data movement from the non participating threads. Since this is within the warp, the data movement is small, so I think it would really need some testing to see if it is actually an optimization, but I see where you are coming from now.

Reply:
performance side. it may or may not be optimization depends on hardware. I'm thinking the  tid 16-31 should be sum = 1; 8-15 should have sum = 2; Looks little either my understanding is wrong or we used it in a wrong way.

Reply:
looks like __shfl_xor and __shfl_down will be deprecated in cuda 9.

Reply:
Yeah, so is __ballot, they are replaced with __*_sync functions, and we can make our code backwards compatible with earlier versions via

```
#if defined(CUDA_VERSION) && CUDA_VERSION < 9000
    #define __ballot_sync(mask, predicate) __ballot(predicate)
    #define __shfl_down_sync(mask, val, delta, width) __shfl_down(val, delta, width)
    #define __shfl_xor_sync(mask, val, laneMask, width) __shfl_xor(val, laneMask, width)
#endif
```
once we move to CUDA 9.

Reply:
yes.